features/mbedtls/mbed_lib.json

```diff
@@ -0,0 +1,157 @@
+{
+    "name": "mbedtls",
+    "config": {
+        "MBED_CLOUD_CLIENT": {
+            "help": "Include support for ciphersuites required by the Mbed Cloud Client",
+            "value": 1
```

Comment:
Turning on the cipher-suites by default for cloud client applications would mean that all other applications who do not require these cipher-suites must explicitly disable these configs in their mbed_app.json.

Reply:
Probably, by default nothing should be turned on. On-demand application can choose to turn on features as required.

Reply:
I see your point to a degree. However, if an developer does not change the default configuration, and does not use any TLS features, the linker should optimize all of MbedTLS out anyway, correct?
I've tried to leave the default settings as close as possible to the current default settings to cause as little disruption to developers as possible.

Reply:
> I've tried to leave the default settings as close as possible to the current default settings to cause as little disruption to developers as possible.
What are the differences ? Could you detail the breaking change made to the default configuration ?

Reply:
@dreemkiller For example LoRaWAN will need AES and CMAC from MbedTLS. However it doesn't require a collection of ciphersuites which are necessary for CLOUD client applications. If we take your current proposal as it is at the moment, the developer has to set MBED_CLOUD_CLIENT = 0 for example. This step wouldn't be needed if we let the appropriate application to choose for itself, i.e., the CLOUD_CLIENT application could set the MBED_CLOUD_CLINET = 1 on demand. Hope you understand what I mean. We can take this offline if you wish.

Reply:
@hasnainvirk You can test that and report the differences between builds, if any

Reply:
A quick test with your PR reveals that the premise of still being backward compatible with the user provided app config file is not true. For example our app config file will undef GCM, CACHAPOLY and POLY1305 and by default mbedtls will turn on pelion-client which will try to define GCM and this will result in compile time error saying the configurations in the two config files are not compatible. I will now experiment with removing all our undefs and then with totally giving away with the app config file and using the config system instead. I will update asap.

Reply:
Update 1: Using App config file even with all the undefs removed
Does not work. Still have the config conflict compile time error.
Here is the app config file I used:
```C
#define MBEDTLS_CIPHER_C
#define MBEDTLS_AES_C
#define MBEDTLS_CMAC_C

// Reduce ROM usage by optimizing some mbedtls features.
// These are only reference configurations for this LoRa example application.
// Other LoRa applications might need different configurations.
#define MBEDTLS_AES_FEWER_TABLES

/*
#undef MBEDTLS_GCM_C
#undef MBEDTLS_CHACHA20_C
#undef MBEDTLS_CHACHAPOLY_C
#undef MBEDTLS_POLY1305_C
*/
```

Reply:
Update 2: Doing away with app config file will not work for us.
The reason is that the config options exposed in the mbed_lib.json for mbedtls doesn't provide us hooks to choose our desired configuration.

Reply:
@dreemkiller I really appreciate what you are trying to do here. This is very important and is absolutely required. But at the moment I am unable make it work with our setup. 
Here is the example application if you wish to do experiments yourself https://github.com/ARMmbed/mbed-os-example-lorawan. Please ping me if you need any assistance with getting up to speed with the application.

Reply:
Thank you, I will take a look at that and try to resolve your problems.

Reply:
@hasnainvirk Thanks so much for this use-case.

From what I can tell, the `mbedtls_lora_config.h` file is not a complete specification of what you need, but just a subset. You were relying on the Mbed TLS configuration file to provide the rest.
This is different from what I expected, so I am very glad you brought this up.

I've submitted a change that takes this into account, and I've compiled your example app with no issues (it required removing the MBEDTLS_USER_CONFIG define from the mbed_app.json file and replacing it with: "mbedtls.app-config-file": "\"mbedtls_lora_config.h\"" in the "*" target_overrides). 

Since I don't have a test set up for that application, I was unable to run it, but it did compile. Could you take a look to ensure that what I've done is consistent with your expectations?

Reply:
@dreemkiller Thanks. I will take your PR and test, and then update you asap.

Reply:
@dreemkiller I have tested your PR with our setup and everything works fine. 
However, I am seeing an increase in flash footprint by 1K (1048 bytes precisely).  
Compiler used is GCC

```
Numbers with your PR:

Elf2Bin: mbed-os-example-lorawan
| Module                     |         .text |    .data |      .bss |
|----------------------------|---------------|----------|-----------|
| [fill]                     |       220(-2) |    4(+0) |  2106(+0) |
| [lib]/c.a                  |     28951(+0) | 2472(+0) |    89(+0) |
| [lib]/gcc.a                |      3232(+0) |    0(+0) |     0(+0) |
| [lib]/m.a                  |       652(+0) |    0(+0) |     0(+0) |
| [lib]/misc                 |       248(+0) |    8(+0) |    28(+0) |
| [lib]/nosys.a              |        32(+0) |    0(+0) |     0(+0) |
| main.o                     |      1828(-4) |    0(+0) |  3892(+0) |
| mbed-lora-radio-drv/SX1272 |       383(+0) |    0(+0) |     0(+0) |
| mbed-lora-radio-drv/SX1276 |      8381(+0) |    0(+0) |     0(+0) |
| mbed-os/cmsis              |      1033(+0) |    0(+0) |     0(+0) |
| mbed-os/components         |       290(+0) |    0(+0) |     0(+0) |
| mbed-os/drivers            |      1532(+0) |    0(+0) |    52(+0) |
| mbed-os/events             |      1490(+0) |    0(+0) |     0(+0) |
| mbed-os/features           |  32252(+1046) |    4(+0) |   184(+0) |
| mbed-os/hal                |      1878(+0) |    8(+0) |   130(+0) |
| mbed-os/platform           |      4310(+0) |  260(+0) |   293(+0) |
| mbed-os/rtos               |     11166(+0) |  168(+0) |  5973(+0) |
| mbed-os/targets            |     13886(+0) |   12(+0) |  1029(+0) |
| trace_helper.o             |         2(+0) |    0(+0) |     0(+0) |
| Subtotals                  | 111766(+1040) | 2936(+0) | 13776(+0) |
Total Static RAM memory (data + bss): 16712(+0) bytes
Total Flash memory (text + data): 114702(+1040) bytes



Numbers with Mbed OS master

Elf2Bin: mbed-os-example-lorawan
| Module                     |         .text |    .data |      .bss |
|----------------------------|---------------|----------|-----------|
| [fill]                     |       218(-2) |    4(+0) |  2106(+0) |
| [lib]/c.a                  |     28951(+0) | 2472(+0) |    89(+0) |
| [lib]/gcc.a                |      3232(+0) |    0(+0) |     0(+0) |
| [lib]/m.a                  |       652(+0) |    0(+0) |     0(+0) |
| [lib]/misc                 |       248(+0) |    8(+0) |    28(+0) |
| [lib]/nosys.a              |        32(+0) |    0(+0) |     0(+0) |
| main.o                     |      1828(+0) |    0(+0) |  3892(+0) |
| mbed-lora-radio-drv/SX1272 |       383(+0) |    0(+0) |     0(+0) |
| mbed-lora-radio-drv/SX1276 |      8381(+0) |    0(+0) |     0(+0) |
| mbed-os/cmsis              |      1033(+0) |    0(+0) |     0(+0) |
| mbed-os/components         |       290(+0) |    0(+0) |     0(+0) |
| mbed-os/drivers            |      1532(+0) |    0(+0) |    52(+0) |
| mbed-os/events             |      1490(+0) |    0(+0) |     0(+0) |
| mbed-os/features           |  31206(-1046) |    4(+0) |   184(+0) |
| mbed-os/hal                |      1878(+0) |    8(+0) |   130(+0) |
| mbed-os/platform           |      4310(+0) |  260(+0) |   293(+0) |
| mbed-os/rtos               |     11166(+0) |  168(+0) |  5973(+0) |
| mbed-os/targets            |     13886(+0) |   12(+0) |  1029(+0) |
| trace_helper.o             |         2(+0) |    0(+0) |     0(+0) |
| Subtotals                  | 110718(-1048) | 2936(+0) | 13776(+0) |
Total Static RAM memory (data + bss): 16712(+0) bytes
Total Flash memory (text + data): 113654(-1048) bytes

``

Reply:
Very interesting.  Thanks for the feedback.

Here's the diagnosis: I coded the `app-config-file` option to be included first, before the rest of the defines. This meant that any `undefs` in the file might get overridden by the rest of the configuration options, including the default ones. 

I have now moved the `app-config-file` option to be last, so it has "supremacy" and can override any of the other options. The size is now smaller than the previous settings (there are a few options that were previously default that have been optimized out when users don't select things).